src/workerd/api/sockets.c++

```diff
@@ -47,122 +47,45 @@ jsg::Ref<Socket> connectImpl(
   return connectImplNoOutputLock(js, kj::mv(actualFetcher), kj::mv(address));
 }
 
-InitData initialiseSocket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) {
+InitData initialiseSocket(
+    jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) {
   auto& context = IoContext::current();
 
-  // Initialise the readable/writable streams with a custom AsyncIoStream that waits for the
+  // Initialise the readable/writable streams with a promised AsyncIoStream that waits for the
   // completion of `connectionPromise` before performing reads/writes.
-  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise));
-  auto sysStreams = newSystemMultiStream(kj::addRef(*stream), StreamEncoding::IDENTITY, context);
+  auto stream = kj::newPromisedStream(kj::mv(connectionPromise));
+  auto sysStreams = newSystemMultiStream(kj::mv(stream), context);
+  auto readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable));
+  auto writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable));
+
+  auto closeFulfiller = IoContext::current().addObject(kj::heap<jsg::PromiseResolverPair<void>>(
+      jsg::newPromiseAndResolver<void>(context.getCurrentLock().getIsolate())));
 
   return {
-    .readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable)),
-    .writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable)),
-    .closeFulfiller = IoContext::current().addObject(
-        kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()))
+    .readable = kj::mv(readable),
+    .writable = kj::mv(writable),
+    .closeFulfiller = kj::mv(closeFulfiller)
   };
 }
 
-Socket::Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
-    Socket(initialiseSocket(kj::mv(connectionPromise))) {};
+Socket::Socket(jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
+    Socket(js, initialiseSocket(js, kj::mv(connectionPromise))) {};
 
-jsg::Promise<void> Socket::close(jsg::Lock& js) {
-  if (!closeFulfiller->fulfiller->isWaiting()) {
-    return js.resolvedPromise();
+void Socket::close() {
```

Comment:
Is it okay if two calls to this function back-to-back both call cancel() and abort()? That is, should we have a concept of a "closing" state?

Reply:
Good question. I will add a TODO comment here to add a test for this.